### PR TITLE
Add query history persistence 

### PR DESCRIPTION
**Pull request by Charles-Gagnon**
Closes https://github.com/microsoft/azuredatastudio/issues/7438

- Query history items are persisted in an encrypted file `queryHistory.bin` - stored in the extension storage location (e.g. `C:\Users\<USER>\AppData\Roaming\azuredatastudio\User\globalStorage\microsoft.query-history`)
- This file is encrypted using the `crypto` library - with the key and IV being generated at startup and then saved to extension secret storage
- Debounce of 10sec is added to file writes to prevent writing the file more than necessary (and then also write file on extension deactivation to flush state)
- If user deactivates persistence then file is deleted

Follow up tasks :
- https://github.com/microsoft/azuredatastudio/issues/7361
- Look into how large query sizes affect this and make sure we handle it as gracefully as we can
- Add tests

### CODE REVIEW

**Comment by cheenamalhotra** [comment]
Overall great concept!

Some feedback on design:
- Can we have single file per query and identify unique queries + target connection before adding them to history ?
- Also we can leverage LruCache npm package to design query cache that identifies least used unique query from the max limit and removes them as needed.

This way we can solve:
- Performance issues writing to and reading from a single file - especially when dealing with large queries and query history grows over time.
- Identifying and auto filtering most used queries to stay above least used.
- Not duplicating unique query history entries but moving them up the list if same query is executed on same connections - this can be huge in saving space consumption and can make room for more elements in history.

File: extensions/query-history/src/queryHistoryProvider.ts [hunk 1]
```diff
@@ -64,28 +84,118 @@ export class QueryHistoryProvider implements vscode.TreeDataProvider<QueryHistor
...
+		try {
+			let ivString = await this._context.secrets.get(STORAGE_IV_KEY);
+			if (!ivString) {
+				iv = crypto.randomBytes(16);
+				await this._context.secrets.store(STORAGE_IV_KEY, iv.toString('binary'));
+			} else {
+				iv = Buffer.from(ivString, 'binary');
+			}
+		} catch (err) {
+			console.error(`Error getting persistance storage IV ${err}`);
```
Comment (alanrenmsft) [inline extensions/query-history/src/queryHistoryProvider.ts:110]: add a `:` before the error message? also applies to other places.

File: extensions/query-history/src/utils.ts [hunk 2]
```diff
@@ -11,3 +11,35 @@
 export function removeNewLines(str: string): string {
 	return str.replace(/\r\n/g, ' ').replace(/\n/g, ' ');
 }
+
+function decorate(decorator: (fn: Function, key: string) => Function): Function {
```
Comment (alanrenmsft) [inline extensions/query-history/src/utils.ts:15]: I suppose you copied this from the `src/vs`, does this one do the job? https://www.npmjs.com/package/debounce-decorator-ts

**Review by alanrenmsft — APPROVED** [review]

File: extensions/query-history/src/utils.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply extensions/query-history/src/utils.ts:15]: Yeah - it's originally from VS (we use it in a lot of extensions). 

For stuff like this which is only a dozen or so lines I usually prefer just to copy over the code. Packages can be problematic and often times have extra stuff we don't care about (plus make it more difficult to actually know what it's doing without spending a bunch of time digging into the source)

**Comment by Charles-Gagnon** [comment]
Thanks for the feedback @cheenamalhotra! I'll look into options when I'm working on adding the max entries, some excellent ideas there.

**Pull request merged**